File: csharp/DesiArchitect.DesignPatterns/06_Adapter/After/DelhiveryAdapter.cs
using DesiArchitect.DesignPatterns.Shared;

namespace DesiArchitect.DesignPatterns.Adapter.After;

public class DelhiveryAdapter : ILogisticsPartner
{
    private readonly DelhiveryClient _client;

    public DelhiveryAdapter(DelhiveryClient client) => _client = client;

    public TrackingId Ship(Parcel parcel)
    {
        Console.WriteLine($"  🔄 DelhiveryAdapter: translating Parcel -> DelhiveryShipmentPayload");
        // Each vendor has a different shape; the adapter absorbs that difference, not the domain.
        var payload = new DelhiveryShipmentPayload(parcel.OrderId, parcel.Address, parcel.WeightKg);
        var response = _client.DispatchShipment(payload);
        return new TrackingId(response.TrackingNumber);
    }
}

File: typescript/src/06-adapter/after/delhivery-adapter.ts
import type { Parcel, TrackingId } from '../../shared/models.js';
import type { ILogisticsPartner } from './ilogistics-partner.js';
import { DelhiveryClient } from './delhivery-client.js';

export class DelhiveryAdapter implements ILogisticsPartner {
  constructor(private readonly client: DelhiveryClient) {}

  ship(parcel: Parcel): TrackingId {
    console.log('  🔄 DelhiveryAdapter: translating Parcel -> DelhiveryShipmentPayload');
    const payload = {
      waybillRef: parcel.orderId,
      dropAddress: parcel.address,
      packageWeight: parcel.weightKg,
    };
    const response = this.client.dispatchShipment(payload);
    return { value: response.trackingNumber };
  }
}